### PR TITLE
ARROW-16692: [C++] StackOverflow in merge generator causes segmentation fault in scan

### PR DESCRIPTION
**Pull request by westonpace**
Fix for the merged generator to avoid a potential stack overflow

### CODE REVIEW

**Comment by westonpace** [comment]
There is some high level description of the problem & the fix in https://issues.apache.org/jira/browse/ARROW-16692

**Review by lidavidm — APPROVED** [review]
The longer term plan is to replace all this with the push-based plan, right? Or does this have to stay for dataset?

**Comment by westonpace** [comment]
> The longer term plan is to replace all this with the push-based plan, right? Or does this have to stay for dataset?

Yes, as part of ARROW-16072.

**Comment by westonpace** [comment]
Hmm, sadly https://github.com/apache/arrow/runs/7477126709?check_suite_focus=true seems relevant.  Perhaps this is leading the merge generator to not properly discard its state.

**Comment by westonpace** [comment]
I investigated the failing ASAN run further today.  I'm able to reproduce it on master so it is unrelated to this change.  The root problem is the same one that motivates ARROW-16072.  From the JIRA:

> We've hit a bit of a wall with the merged generator. The current behavior is: If one subscription encounters an error we simply stop pulling from the other subscriptions. Once everything has settled down we return the error and end the stream.
> 
> In reality, we should be sending some kind of cancel signal down to the other generators. Otherwise, we are not respecting the rule for AsyncGenerator that we recently defined which is "An AsyncGenerator should always be fully consumed".

The end result of this is that some bits of the scanner keep running in the event of a failure or an abort.  These pieces keep their state alive via shared_ptr so it doesn't actually cause any exceptions.  However, it can lead to cases like this memory leak.

The last test in ScannerTest will test a failed scan.  This causes some bits to be left running.  Then the test ends and the process shuts down and the thread pool shuts down immediately.  This causes some cleanup thread tasks to fail to get scheduled or not run.  As a result there is some leakage at shutdown.

This is only a shutdown leak and so pretty minor.  ARROW-16072 will tie cancellation into the scanner so that once a plan is aborted and the user waits for it to finish then nothing will be left running and there will be no possibility for this.

Long story short, I think we can proceed with merging this PR.

**Comment by westonpace** [comment]
I've filed ARROW-17198 for the memory leak.

**Comment by lidavidm** [comment]
Can we add an ASAN suppression temporarily before merging?

**Comment by lidavidm** [comment]
(Also probably worth doing the same for Valgrind)

**Comment by westonpace** [comment]
> Can we add an ASAN suppression temporarily before merging?

I added a `WaitForIdle` at the end of the test.  This should prevent any possible leak, from this test at least.

**Review by lidavidm — APPROVED** [review]

**Comment by westonpace** [comment]
Failure seems unrelated.  I'm going to merge.
